### ISSUE TITLE
@!method tags are not included in --embed-mixins

### ISSUE THREAD

**Issue by aaronmallen** [lsegal/yard#1305]
[Describe your issue here. Try to be as specific as possible so we can find
the root cause more quickly!]

`@!method` tags are not included in `--embed-mixins`

## Steps to reproduce

This is the minimal reproduction for the issue. I've done my best to remove
all extraneous code and unique environment state on my machine before providing
these steps:
1. create a module with a method tag and a class that includes it:
```ruby
module A
  # @!method foo
end

class B
  include A
end
```
2. run `yard doc --embed-mixins`
3. Goto the generated documentation for `B`

## Expected Output
methods defined with the `@!method` tag should be included with other mixins

[Describe what you *expected* to happen.]

## Environment details:

* OS: MacOS Catalina v 10.15.2
* Ruby version (`ruby -v`): ruby 2.5.0p0 (2017-12-25 revision 61468) [x86_64-darwin18]
* YARD version (`yard -v`): yard 0.9.24

I have read the [Contributing Guide][contrib].

[contrib]: https://github.com/lsegal/yard/blob/master/CONTRIBUTING.md

**Comment by lsegal** [comment]
It looks like there are 2 issues with this reproduction case:

1. Firstly, the reason you're seeing no method in B is that no method is actually being defined here. If you look at the docs for A, you will see there is no method defined in A, either. Directives need to be attached to code blocks and cannot end a block, so you would either need to do:

```ruby
class A
  # @!method foo
  some_dsl_method_to_attach_to
end
```

Or attach it to the outer class:

```ruby
# @!method foo
#   Attached to the A class.
class A; end
```

2. Secondly, in your example, A is a class, so it cannot be legally `include`d into B. If you change it to `module` alongside the docstring change in (1), the method correctly shows in B when you run `yard doc --embed-mixins`.

In other words, this works:

```ruby
# @!method foo
module A; end

class B
  include A
end
```

Hope that helps!

**Comment by aaronmallen** [comment]
@isegal I'll correct the step for reproduction as that was an obvious error however on your first point I have 2 issues with believing that is the truth:

1. #foo shows up in the documentation for `A`
2. In my actual code I am calling this on top of a `delegate` definition.

**Comment by lsegal** [comment]
> #foo shows up in the documentation for A

I'm unable to reproduce this behavior with the minimal repro provided. Perhaps you can add in some of the syntax regarding the `delegate` syntax?

**Comment by aaronmallen** [comment]
Here is a real world reproduction case:

```ruby
module Context
    # @!method context_fail!(errors = nil)
    #   @see Base::Context#fail!
    def self.included(base)
      delegate :fail!, to: :context, prefix: true
    end
end

class Base
  include Context
end
```

The instances methods for `Context`:
![Screen Shot 2020-01-24 at 12 33 33 AM](https://user-images.githubusercontent.com/931094/73049148-38810f80-3e41-11ea-8a4d-bd9df0a5aee4.png)

The instance methods for `Base`:
![Screen Shot 2020-01-24 at 12 34 13 AM](https://user-images.githubusercontent.com/931094/73049175-5189c080-3e41-11ea-9fd5-5eda63f544d4.png)

**Comment by aaronmallen** [comment]
https://github.com/aaronmallen/activeinteractor/commit/f1fc28ec7caef920caf681a579b07afc16bb6261 here is the code in it's entirety

**Comment by lsegal** [comment]
I'm still a little unclear on the expected result. I see `#context_fail!` included correctly on Base when I run `yard --embed-mixins` with the snippet provided in the above comment:

![image](https://user-images.githubusercontent.com/686/73049460-8726ad80-3e31-11ea-9b56-4569200aaff0.png)

This looks correct to me. Perhaps something else needs to be introduced for proper reproduction?

**Comment by aaronmallen** [comment]
I tried directly from the command line with `yard --embed-mixins` and still reproduced the error.  Here is my `.yardopts` file:

```
--title ActiveInteractor
--readme README.md
--markup-provider redcarpet
--markup markdown
--exclude lib/rails
--protected
--no-private
--embed-mixins
--output-dir doc/public
lib/**/*.rb
-
CHANGELOG.md
LICENSE
```

**Comment by lsegal** [comment]
Can you provide the output from `yard --embed-mixins --debug`?

**Comment by aaronmallen** [comment]
debug:
```
[debug]: Parsing ["lib/**/*.rb"] with `ruby` parser
[debug]: Parsing lib/active_interactor.rb
[debug]: Parsing lib/active_interactor/base.rb
[debug]: Parsing lib/active_interactor/error.rb
[debug]: Parsing lib/active_interactor/rails.rb
[debug]: Parsing lib/active_interactor/config.rb
[debug]: Parsing lib/active_interactor/version.rb
[debug]: Parsing lib/active_interactor/configurable.rb
[debug]: Parsing lib/active_interactor/context/base.rb
[debug]: Parsing lib/active_interactor/rails/models.rb
[debug]: Parsing lib/active_interactor/rails/railtie.rb
[debug]: Parsing lib/active_interactor/context/loader.rb
[debug]: Parsing lib/active_interactor/context/status.rb
[debug]: Parsing lib/active_interactor/organizer/base.rb
[debug]: Parsing lib/active_interactor/interactor/worker.rb
[debug]: Parsing lib/active_interactor/organizer/perform.rb
[debug]: Parsing lib/active_interactor/context/attributes.rb
[debug]: Parsing lib/active_interactor/interactor/context.rb
[debug]: Parsing lib/active_interactor/interactor/perform.rb
[debug]: Parsing lib/active_interactor/organizer/organize.rb
[debug]: Parsing lib/active_interactor/organizer/callbacks.rb
[debug]: Parsing lib/active_interactor/interactor/callbacks.rb
[debug]: Parsing lib/active_interactor/rails/orm/active_record.rb
[debug]: Parsing lib/active_interactor/organizer/interactor_interface.rb
[debug]: Parsing lib/active_interactor/organizer/interactor_interface_collection.rb
[debug]: Serializing to .yardoc/objects/root.dat
[debug]: Re-generating object ...
[debug]: Re-generating object ActiveInteractor...
[debug]: Re-generating object ActiveInteractor::Context...
[debug]: Re-generating object ActiveInteractor::Interactor...
[debug]: Re-generating object ActiveInteractor::Organizer...
[debug]: Re-generating object ActiveInteractor::Error...
[debug]: Re-generating object ActiveInteractor::Configurable...
[debug]: Re-generating object ActiveInteractor::Configurable::ClassMethods...
[debug]: Re-generating object ActiveInteractor::Rails...
[debug]: Re-generating object ActiveInteractor::Rails::Models...
[debug]: Re-generating object ActiveInteractor::Rails::Models::ClassMethods...
[debug]: Re-generating object ActiveInteractor::Rails::Models::InstanceMethods...
[debug]: Re-generating object ActiveInteractor::Context::Loader...
[debug]: Re-generating object ActiveInteractor::Context::Status...
[debug]: Re-generating object ActiveInteractor::Organizer::Perform...
[debug]: Re-generating object ActiveInteractor::Organizer::Perform::ClassMethods...
[debug]: Re-generating object ActiveInteractor::Context::Attributes...
[debug]: Re-generating object ActiveInteractor::Context::Attributes::ClassMethods...
[debug]: Re-generating object ActiveInteractor::Interactor::Context...
[debug]: Re-generating object ActiveInteractor::Interactor::Context::ClassMethods...
[debug]: Re-generating object ActiveInteractor::Interactor::Perform...
[debug]: Re-generating object ActiveInteractor::Interactor::Perform::ClassMethods...
[debug]: Re-generating object ActiveInteractor::Organizer::Organize...
[debug]: Re-generating object ActiveInteractor::Organizer::Organize::ClassMethods...
[debug]: Re-generating object ActiveInteractor::Organizer::Callbacks...
[debug]: Re-generating object ActiveInteractor::Organizer::Callbacks::ClassMethods...
[debug]: Re-generating object ActiveInteractor::Interactor::Callbacks...
[debug]: Re-generating object ActiveInteractor::Interactor::Callbacks::ClassMethods...
[debug]: Re-generating object ActiveInteractor::Base...
[debug]: Re-generating object ActiveInteractor::Error::ContextFailure...
[debug]: Re-generating object ActiveInteractor::Error::InvalidContextClass...
[debug]: Re-generating object ActiveInteractor::Config...
[debug]: Re-generating object ActiveInteractor::Context::Base...
[debug]: Re-generating object ActiveInteractor::Rails::Railtie...
[debug]: Re-generating object ActiveInteractor::Organizer::Base...
[debug]: Re-generating object ActiveInteractor::Interactor::Worker...
[debug]: Re-generating object ActiveInteractor::Interactor::Perform::Options...
[debug]: Re-generating object ActiveInteractor::Organizer::InteractorInterface...
[debug]: Re-generating object ActiveInteractor::Organizer::InteractorInterfaceCollection...
[debug]: Generating asset js/jquery.js
[debug]: Serializing to doc/public/js/jquery.js
[debug]: Generating asset js/app.js
[debug]: Serializing to doc/public/js/app.js
[debug]: Generating asset js/full_list.js
[debug]: Serializing to doc/public/js/full_list.js
[debug]: Generating asset css/style.css
[debug]: Serializing to doc/public/css/style.css
[debug]: Generating asset css/common.css
[debug]: Serializing to doc/public/css/common.css
[debug]: Generating asset css/full_list.css
[debug]: Serializing to doc/public/css/full_list.css
[debug]: Generating asset class_list.html
[debug]: Serializing to doc/public/class_list.html
[debug]: Generating asset method_list.html
[debug]: Serializing to doc/public/method_list.html
[debug]: Generating asset file_list.html
[debug]: Serializing to doc/public/file_list.html
[debug]: Generating asset frames.html
[debug]: Serializing to doc/public/frames.html
[debug]: Serializing to doc/public/_index.html
[debug]: Serializing to doc/public/index.html
[debug]: Serializing to doc/public/file.README.html
[debug]: Serializing to doc/public/file.CHANGELOG.html
[debug]: Serializing to doc/public/file.LICENSE.html
[debug]: Serializing to doc/public/top-level-namespace.html
[debug]: Serializing to doc/public/ActiveInteractor.html
[debug]: Serializing to doc/public/ActiveInteractor/Context.html
[debug]: Serializing to doc/public/ActiveInteractor/Interactor.html
[debug]: Serializing to doc/public/ActiveInteractor/Organizer.html
[debug]: Serializing to doc/public/ActiveInteractor/Error.html
[debug]: Serializing to doc/public/ActiveInteractor/Configurable.html
[debug]: Serializing to doc/public/ActiveInteractor/Configurable/ClassMethods.html
[debug]: Serializing to doc/public/ActiveInteractor/Rails.html
[debug]: Serializing to doc/public/ActiveInteractor/Rails/Models.html
[debug]: Serializing to doc/public/ActiveInteractor/Rails/Models/ClassMethods.html
[debug]: Serializing to doc/public/ActiveInteractor/Rails/Models/InstanceMethods.html
[debug]: Serializing to doc/public/ActiveInteractor/Context/Loader.html
[debug]: Serializing to doc/public/ActiveInteractor/Context/Status.html
[debug]: Serializing to doc/public/ActiveInteractor/Organizer/Perform.html
[debug]: Serializing to doc/public/ActiveInteractor/Organizer/Perform/ClassMethods.html
[debug]: Serializing to doc/public/ActiveInteractor/Context/Attributes.html
[debug]: Serializing to doc/public/ActiveInteractor/Context/Attributes/ClassMethods.html
[debug]: Serializing to doc/public/ActiveInteractor/Interactor/Context.html
[debug]: Serializing to doc/public/ActiveInteractor/Interactor/Context/ClassMethods.html
[debug]: Serializing to doc/public/ActiveInteractor/Interactor/Perform.html
[debug]: Serializing to doc/public/ActiveInteractor/Interactor/Perform/ClassMethods.html
[debug]: Serializing to doc/public/ActiveInteractor/Organizer/Organize.html
[debug]: Serializing to doc/public/ActiveInteractor/Organizer/Organize/ClassMethods.html
[debug]: Serializing to doc/public/ActiveInteractor/Organizer/Callbacks.html
[debug]: Serializing to doc/public/ActiveInteractor/Organizer/Callbacks/ClassMethods.html
[debug]: Serializing to doc/public/ActiveInteractor/Interactor/Callbacks.html
[debug]: Serializing to doc/public/ActiveInteractor/Interactor/Callbacks/ClassMethods.html
[debug]: Serializing to doc/public/ActiveInteractor/Base.html
[debug]: Serializing to doc/public/ActiveInteractor/Error/ContextFailure.html
[debug]: Serializing to doc/public/ActiveInteractor/Error/InvalidContextClass.html
[debug]: Serializing to doc/public/ActiveInteractor/Config.html
[debug]: Serializing to doc/public/ActiveInteractor/Context/Base.html
[debug]: Serializing to doc/public/ActiveInteractor/Rails/Railtie.html
[debug]: Serializing to doc/public/ActiveInteractor/Organizer/Base.html
[debug]: Serializing to doc/public/ActiveInteractor/Interactor/Worker.html
[debug]: Serializing to doc/public/ActiveInteractor/Interactor/Perform/Options.html
[debug]: Serializing to doc/public/ActiveInteractor/Organizer/InteractorInterface.html
[debug]: Serializing to doc/public/ActiveInteractor/Organizer/InteractorInterfaceCollection.html
Files:          22
Modules:        27 (   18 undocumented)
Classes:        11 (   10 undocumented)
Constants:       4 (    2 undocumented)
Attributes:      6 (    0 undocumented)
Methods:        60 (   26 undocumented)
 48.15% documented
```

I have checked in the actual code that reproduces here (with slight adjustments but still reproduced):

https://github.com/aaronmallen/activeinteractor/blob/chore/refactor/lib/active_interactor/interactor/context.rb#L37 

and this is the including class here: https://github.com/aaronmallen/activeinteractor/blob/chore/refactor/lib/active_interactor/base.rb#L11

**Comment by lsegal** [comment]
Oh ok so we're not using the minimal reproduction case, but rather your library in its entirety. That makes it a bit hard to pin down the issue.

Without more information about what behavior you're seeing and expecting regarding that codebase, it would be easier to stick to minimal cases.

Specifically, this code provided works correctly for me:

```ruby
module Context
  # @!method context_fail!(errors = nil)
  #   @see Base::Context#fail!
  def self.included(base)
    delegate :fail!, to: :context, prefix: true
  end
end

class Base
  include Context
end
```

Are you getting different results with _this_ provided minimal repro? If so, what are your expected / actual results?

**Comment by aaronmallen** [comment]
nope you are correct with the simple use case I am unable to reproduce 🤔

**Comment by aaronmallen** [comment]
ah so this is interesting:
![Screen Shot 2020-01-24 at 1 06 21 AM](https://user-images.githubusercontent.com/931094/73050520-c65ef980-3e45-11ea-88ae-cd4c07e81c7d.png)

`Interactor::Context::ClassMethods` is not linked in the extended by and `Interactor::Context` isn't even included in the includes list... I'll see if I can trouble shoot further and generate an actual reproduction case.  Thanks for your help @lsegal

**Comment by aaronmallen** [comment]
@lsegal just FYI I don't know if you want me to open a new bug on this or not but the issue is I have `ActiveInteractor::Context` and `ActiveInteractor::Interactor::Context` in `ActiveInteractor::Base` I was including and extending with like this:

```ruby
module ActiveInteractor
  class Base
     extend Interactor::Context::ClassMethods
     include Interactor::Context
  end
end
```

I was able to fix the issue by using the entire module namespace:
```ruby
module ActiveInteractor
  class Base
     extend ActiveInteractor::Interactor::Context::ClassMethods
     include ActiveInteractor::Interactor::Context
  end
end
```

If this is the intended behavior I'll move on with my work around, but if I need to open a new bug let me know.

**Comment by lsegal** [comment]
@aaronmallen that looks like it might be a namespace lookup bug. `ActiveInteractor::Interactor::Context` shouldn't resolve to `ActiveInteractor::Context` unless `ActiveInteractor` included `ActiveInteractor::Interactor` somewhere and the Context inclusion was parsed before that latter include was processed. If you can find a minimal repro case of that setup without any cross-module inclusion, that seems like a legitimate bug.
